### PR TITLE
[EGD-8203] Add URC counter and csq poll mode

### PR DESCRIPTION
**Pull request by kkleczkowski**
Added CSQ notifications counter and CSQ poll mode.
It will help to reduce current cosnuption in case of
variable signal.

### CODE REVIEW

File: module-cellular/at/cmd/src/CSQ.cpp [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+            }
+
+            return p;
+        }
+
+    } // namespace cmd
+
+    namespace result
+    {
+        CSQ::CSQ(const Result &that) : Result(that)
```
Comment (mjan-dev) [inline module-cellular/at/cmd/src/CSQ.cpp:73]:
```suggestion
        explicit CSQ::CSQ(const Result &that) : Result(that)
```

File: module-cellular/at/cmd/src/CSQ.cpp [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+                if (tokens.size() != csq::tokensCount) {
+                    LOG_ERROR("Can't parse - wrong token count");
+                    p.code = result::CSQ::Code::PARSING_ERROR;
+                    return p;
+                }
+
+                auto csq = 0;
+                auto ber = 0;
+                if (utils::toNumeric(tokens[magic_enum::enum_integer(csq::Tokens::Csq)], csq) &&
+                    utils::toNumeric(tokens[magic_enum::enum_integer(csq::Tokens::Ber)], ber)) {
```
Comment (SP2FET) [inline module-cellular/at/cmd/src/CSQ.cpp:56]: consider using `.at()` instead of `[]`

**Review by SP2FET — APPROVED** [review]

File: module-cellular/at/cmd/src/CSQ.cpp (hunk 2 above)
Comment (kkleczkowski) [reply module-cellular/at/cmd/src/CSQ.cpp]: done

File: module-cellular/at/cmd/src/CSQ.cpp [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+#include <type_traits>
+#include <at/cmd/CSQ.hpp>
+#include <chrono>
+
+namespace at
+{
+    namespace cmd
+    {
+        using namespace std::chrono_literals;
+        CSQ::CSQ(at::cmd::Modifier mod) noexcept : Cmd("AT+CSQ", mod, 17s)
```
[the remark this answers is not in the document]
Comment (kkleczkowski) [reply module-cellular/at/cmd/src/CSQ.cpp]: done

File: module-cellular/at/cmd/src/CSQ.cpp (hunk 1 above)
Comment (kkleczkowski) [reply module-cellular/at/cmd/src/CSQ.cpp:73]: done

File: module-services/service-cellular/src/ServiceCellularPriv.cpp [hunk 4]
```diff
@@ -390,4 +394,98 @@ namespace cellular::internal
             return false;
         };
     }
+
+    void ServiceCellularPriv::initCSQHandler()
+    {
+        csqHandler->onEnableCsqReporting = [this]() {
+            auto handle = owner->getTaskHandle();
+            if (owner->cpuSentinel) {
+                owner->cpuSentinel->HoldMinimumFrequencyAndWait(bsp::CpuFrequencyMHz::Level_4, handle, 2000);
```
Comment (pholat) [inline module-services/service-cellular/src/ServiceCellularPriv.cpp:403]: it's done so that next channel->cmd wouldn't fail?

**Review by pholat — APPROVED** [review]

File: module-services/service-cellular/service-cellular/CellularMessage.hpp [hunk 5]
```diff
@@ -1018,4 +1018,38 @@ namespace cellular
       public:
         RetryPhoneModeChangeRequest() : sys::DataMessage(MessageType::MessageTypeUninitialized){};
     };
+
+    class URCCounterMessage : public sys::DataMessage
+    {
+      public:
+        URCCounterMessage(const uint32_t urcCounter)
```
Comment (mjan-dev) [inline module-services/service-cellular/service-cellular/CellularMessage.hpp:1025]: explicit

File: module-services/service-cellular/src/CSQHandler.cpp [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+#include <log/log.hpp>
+
+namespace cellular::service
+{
+
+    void CSQHandler::handleTimerTick()
+    {
+        if (isInPollMode()) {
+            timeSpendInPollMode++;
```
Comment (mjan-dev) [inline module-services/service-cellular/src/CSQHandler.cpp:14]: minor typo 
```suggestion
            timeSpentInPollMode++;
```

File: module-services/service-cellular/src/ServiceCellularPriv.cpp (hunk 4 above)
Comment (mjan-dev) [inline module-services/service-cellular/src/ServiceCellularPriv.cpp:403]: Maybe we could move the magic value to constexpr

File: module-services/service-cellular/src/CSQHandler.cpp [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+                auto csq = result.value();
+                if (csq.csq != invalid_rssi_low && csq.csq != invalid_rssi_high) {
+                    LOG_INFO("Propagate valid CSQ");
+                    if (onPropagateCSQ != nullptr) {
+                        onPropagateCSQ(csq.csq);
+                        return true;
+                    }
+                }
+                else {
+                    LOG_INFO("Inavalid CSQ, notify service antenna");
```
Comment (mjan-dev) [inline module-services/service-cellular/src/CSQHandler.cpp:91]:
```suggestion
                    LOG_INFO("Invalid CSQ, notify service antenna");
```

File: module-services/service-cellular/src/ServiceCellularPriv.cpp (hunk 4 above)
Comment (kkleczkowski) [reply module-services/service-cellular/src/ServiceCellularPriv.cpp:403]: yes

File: module-services/service-cellular/src/CSQHandler.hpp [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+        std::function<void()> onRetryGetCSQ;
+
+        bool switchToReportMode();
+        bool switchToPollMode();
+        bool getCSQ();
+
+      private:
+        uint32_t urcCounter = 0;
+        CSQMode currentMode = CSQMode::Reporting;
+        std::chrono::minutes timeSpendInPollMode;
```
Comment (mjan-dev) [inline module-services/service-cellular/src/CSQHandler.hpp:49]:
```suggestion
        std::chrono::minutes timeSpendInPollMode{};
```

**Review by mjan-dev — APPROVED** [review]
Added minor comments. As far as logic goes it seems alright to me.

File: module-services/service-cellular/service-cellular/CellularMessage.hpp (hunk 5 above)
Comment (kkleczkowski) [reply module-services/service-cellular/service-cellular/CellularMessage.hpp:1025]: done

File: module-services/service-cellular/src/CSQHandler.cpp (hunk 6 above)
Comment (kkleczkowski) [reply module-services/service-cellular/src/CSQHandler.cpp:14]: done

File: module-services/service-cellular/src/ServiceCellularPriv.cpp (hunk 4 above)
Comment (kkleczkowski) [reply module-services/service-cellular/src/ServiceCellularPriv.cpp:403]: done

File: module-services/service-cellular/src/CSQHandler.cpp (hunk 7 above)
Comment (kkleczkowski) [reply module-services/service-cellular/src/CSQHandler.cpp:91]: done

File: module-services/service-cellular/src/CSQHandler.hpp (hunk 8 above)
Comment (kkleczkowski) [reply module-services/service-cellular/src/CSQHandler.hpp:49]: done

**Pull request merged**